File: 1/2021.04.07/710.Random Pick with Blacklist/java/solution.java
class Solution {

    Map<Integer, Integer> map;
    Random random;
    int wlen;
    
    public Solution(int N, int[] blacklist) {
        map = new HashMap<>();
        random = new Random();
        wlen = N - blacklist.length;
        Set<Integer> w = new HashSet<>();
        for (int i = wlen; i < N; i++) {
            w.add(i);
        }
        for (int i : blacklist) {
            w.remove(i);
        }
        Iterator<Integer> iterator = w.iterator();
        for (int i : blacklist) {
            if (i < wlen) {
                map.put(i, iterator.next());
            }
        }
    }
    
    public int pick() {
        int k = random.nextInt(wlen);
        return map.getOrDefault(k, k);
    }
}

File: 1/2021.04.07/710.Random Pick with Blacklist/python/solution.py
class Solution:

    def __init__(self, N: int, blacklist: List[int]):
        wlen = N - len(blacklist)
        w = set()
        for i in range(wlen, N):
            w.add(i)
        for i in blacklist:
            if i in w:
                w.remove(i)
        it = iter(w)
        self.map = {}
        for i in blacklist:
            if i < wlen:
                self.map[i] = next(it)
        self.wlen = wlen
        

    def pick(self) -> int:
        k = random.randint(0, self.wlen - 1)
        return self.map.get(k, k)
